#### ETAPE 1 : CHOIX DU SECTEUR

#### ETAPE 2 : TROUVER UNE PROBLEMATIQUE - IDENTIFIER UNE TARGET

#### ETAPE 3 : SELECTIONNER UNE OU DES BASES DE DONNEES (SQL, CSV, API, SCRAPING, ETC...).

#### ETAPE 4 : REALISER L'ANALYSE DE DONNEES

#### ETAPE 5 : REALISER  UN MODEL DE MACHINE LEARNING (REGRESSION LINEAIRE)

#### ETAPE 6 : FAIRE UNE DEMO API ( PREDICATION - FAST API)

1 - CHOIX DU SECTEUR :
    
    L'ENVIRONNEMENT ET LES CONSEQUENCES DE LA POLLUTION DE L'AIR SUR LA POPULATION MONDIALE

2 - PROBLEMATIQUE ET IDENTIFICATION DE LA TARGET :
    
    LA POLLUTION DE L'AIR A T IL UNE INFLUENCE FORTE SUR LE TAUX DE MORTALITE SUR LA POPULATION MONDIALE ?
    
    TARGET : POLLUTION LIE A L'OZONE
    
   DEF : La pollution de l'air est un ensemble de gaz et de particules en suspension présents dans l'air (intérieur ou extérieur) dont les niveaux de concentration varient en fonction des émissions et des conditions météorologiques, et qui sont nuisibles pour la santé et l'environnement

Rappels : 
-   : 10 minutes max. Répartis de cette manière : 
5 minutes présentation + 3 minutes de questions / réponses + 2 minutes demo API.
-   : Pause de 5 minutes entre 2 interventions. 
-   : Pause à 15h30 jeudi  et à 10h30 vendredi.

  Mindset 
C'est l'explication de la "démarche" qui est intéressante  , plutôt que de montrer des lignes de code  .
L'explication de la "démarche" s'appuie notamment sur :
-   : L'interprétation des graphiques. 
-   : L'explication du choix de la target, des features, et du model .
-   : L'explication de la méthode d'entrainement du model (exemple 1 : hold out methode. Exemple 2 : itération_1 --> Base line avec une feature, le score est à 0,5 . itération_2 -->  model avec 2 features le score est à 0,6. Etc...)
-   : L'explication du nettoyage des données. (exemple : j'ai remplacé les valeurs manquantes par la médiane parceque ...)
-   : L'explication de la metrique de performance (du score) : (https://scikit-learn.org/stable/modules/model_evaluation.html) 
-   : L'explication du choix de l'architecture du code. 

I - CHOIX DU DATASET VIA KAGGLE : ETUDE SUR LE TAUX DE MORTALITE CAUSE PAR LA POLLUTION DE L'AIR

II - IMPORT DE MODULES NECESSAIRE AU TRAVAIL A REALISER

In [1]:
import numpy as np
import pandas as pd
import requests
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
import matplotlib
import matplotlib.pyplot as plt
from matplotlib.patches import ConnectionPatch
from matplotlib.patches import Circle
import matplotlib.pyplot as plt
import plotly.figure_factory as ff
import plotly.graph_objects as go
import plotly.express as px
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_validate, cross_val_score, train_test_split, GridSearchCV


In [2]:
df = pd.read_csv("death-rates-from-air-pollution.csv")


III - ANALYSE DU DATASET

In [3]:
df.head() # Visualisation globale du Dataset

Entity Code  Year  Air pollution (total) (deaths per 100,000)  \
0  Afghanistan  AFG  1990                                  299.477309   
1  Afghanistan  AFG  1991                                  291.277967   
2  Afghanistan  AFG  1992                                  278.963056   
3  Afghanistan  AFG  1993                                  278.790815   
4  Afghanistan  AFG  1994                                  287.162923   

   Indoor air pollution (deaths per 100,000)  \
0                                 250.362910   
1                                 242.575125   
2                                 232.043878   
3                                 231.648134   
4                                 238.837177   

   Outdoor particulate matter (deaths per 100,000)  \
0                                        46.446589   
1                                        46.033841   
2                                        44.243766   
3                                        44.440148   
4                                        45.594328   

   Outdoor ozone pollution (deaths per 100,000)  
0                                      5.616442  
1                                      5.603960  
2                                      5.611822  
3                                      5.655266  
4                                      5.718922

In [4]:
df.isna().any()  # isna() : Détection de valeurs manquantes

Entity                                             False
Code                                                True
Year                                               False
Air pollution (total) (deaths per 100,000)         False
Indoor air pollution (deaths per 100,000)          False
Outdoor particulate matter (deaths per 100,000)    False
Outdoor ozone pollution (deaths per 100,000)       False
dtype: bool

In [5]:
df.isna().sum() # 980 valeurs manquantes sur la colonne : 'Code'

Entity                                               0
Code                                               980
Year                                                 0
Air pollution (total) (deaths per 100,000)           0
Indoor air pollution (deaths per 100,000)            0
Outdoor particulate matter (deaths per 100,000)      0
Outdoor ozone pollution (deaths per 100,000)         0
dtype: int64

In [6]:
df['Year'].unique() # Verification des données unique sur la colonne 'Year'

array([1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015, 2016, 2017])

In [7]:
df['Entity'].nunique() # Vérification du nombre de 'pays' sur la DATASET

231

In [8]:
df['Code'].nunique() # Vérification du nombre de 'code' sur la DATASET

196

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6468 entries, 0 to 6467
Data columns (total 7 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   Entity                                           6468 non-null   object 
 1   Code                                             5488 non-null   object 
 2   Year                                             6468 non-null   int64  
 3   Air pollution (total) (deaths per 100,000)       6468 non-null   float64
 4   Indoor air pollution (deaths per 100,000)        6468 non-null   float64
 5   Outdoor particulate matter (deaths per 100,000)  6468 non-null   float64
 6   Outdoor ozone pollution (deaths per 100,000)     6468 non-null   float64
dtypes: float64(4), int64(1), object(2)
memory usage: 353.8+ KB


In [66]:
df.describe()

Year    total_air  indoor_pollution  outdoor_particule  \
count  6468.000000  6468.000000       6468.000000        6468.000000   
mean   2003.500000    87.804083         50.481504          35.114982   
std       8.078372    57.845436         58.722565          18.207369   
min    1990.000000     8.401696          0.003859           7.542899   
25%    1996.750000    41.016114          2.002779          21.955108   
50%    2003.500000    72.653557         23.737851          30.977196   
75%    2010.250000   127.607177         90.138777          44.604181   
max    2017.000000   318.463253        293.509187         132.815726   

       outdoor_ozone  
count    6468.000000  
mean        3.370793  
std         4.325445  
min         0.000000  
25%         1.302905  
50%         2.338684  
75%         3.515409  
max        37.356034

CONCLUSION

LE DATASET EST COMPOSE DE 7 COLONNES REPRESENTANT :

    - LE NOM DU PAYS OU ZONE ( 231 PAYS )
    
    - LE CODE PAYS
    
    - L'ANNEE  ( DE 1990 à 2017 )
    
    - LE NOMBRE DE MORTS TOTAL CAUSES PAR LA POLLUTION DE L'AIR (MORTS/100 000)
    
    - LE NOMBRE DE MORTS CAUSES PAR LA POLLUTION DE L'AIR INTERIEUR (MORTS/100 000)
    
    - LE NOMBRE DE MORTS CAUSES PAR LA POLLUTION DE L'AIR DES PARTICULES EN SUSPENSION (MORTS/100 000)
    
    - LE NOMBRE DE MORTS CAUSES PAR LA POLLUTION ATMOSPHERIQUE A L'OZONE (MORTS/100 000)




    
1- POLLUTION DE L'AIR DES PARTICULES EN SUSPENSION : 
La toxicité des particules en suspension est essentiellement due aux particules de diamètre inférieur à 10µm. Elles peuvent être émises directement dans l'air par des activités anthropiques (industrie, résidentiel, agriculture, transports) et par des sources naturelles (feux de forêt, éruptions volcaniques, etc.).

2- POLLUTION ATMOSPHERIQUE A L'OZONE :
L’ozone est un polluant atmosphérique bien connu pour ses effets délétères sur la santé humaine et sur la végétation. Il se forme dans les basses couches de l’atmosphère en période estivale, sous l’effet du rayonnement solaire et lorsque les températures sont élevées

3- POLLUTION DE L'AIR INTERIEUR :
Les sources de pollution dans les logements sont nombreuses : tabagisme, moisissures, matériaux de construction, meubles, acariens, produits d'entretien, peintures...

 IV - NETTOYAGE DU DATASET

In [11]:
df.rename(columns = {'Air pollution (total) (deaths per 100,000)' : 'total_air', 'Entity' : 'country','Indoor air pollution (deaths per 100,000)' : 'indoor_pollution','Outdoor particulate matter (deaths per 100,000)': 'outdoor_particule', 'Outdoor ozone pollution (deaths per 100,000)' : 'outdoor_ozone'}, inplace = True)
# Renommage des colonnes pour l'entrainement du modèle

In [12]:
df

country Code  Year   total_air  indoor_pollution  outdoor_particule  \
0     Afghanistan  AFG  1990  299.477309        250.362910          46.446589   
1     Afghanistan  AFG  1991  291.277967        242.575125          46.033841   
2     Afghanistan  AFG  1992  278.963056        232.043878          44.243766   
3     Afghanistan  AFG  1993  278.790815        231.648134          44.440148   
4     Afghanistan  AFG  1994  287.162923        238.837177          45.594328   
...           ...  ...   ...         ...               ...                ...   
6463     Zimbabwe  ZWE  2013  143.850145        113.456097          27.589603   
6464     Zimbabwe  ZWE  2014  138.200536        108.703566          26.760618   
6465     Zimbabwe  ZWE  2015  132.752553        104.340506          25.715415   
6466     Zimbabwe  ZWE  2016  128.692138        100.392287          25.643570   
6467     Zimbabwe  ZWE  2017  125.028843         96.235996          26.166182   

      outdoor_ozone  
0          5.616442  
1          5.603960  
2          5.611822  
3          5.655266  
4          5.718922  
...             ...  
6463       4.426291  
6464       4.296971  
6465       4.200907  
6466       4.117173  
6467       4.052495  

[6468 rows x 7 columns]

#### ENTRAINEMENENT DU MODELE : TRAIN_TEST_SPLIT

In [13]:
from sklearn.model_selection import train_test_split

In [36]:
y = df['outdoor_ozone']
X = df.drop(['outdoor_ozone','country','Code','Year'], axis = 1)

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,random_state=42)

#### CREATION DU PIPELINE : ONEHOTENCODER/MINMAXSCALER/LINEARREGRESSION

In [38]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer

In [39]:
from sklearn.compose import ColumnTransformer

In [40]:
from sklearn.pipeline import make_pipeline

In [41]:
from sklearn import set_config
set_config(display="diagram")

In [42]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [43]:
#num_pipeline = Pipeline([
    #('imputer', SimpleImputer(strategy="median")),
    #('minmax_scaler', MinMaxScaler())
#])
#ohe = OneHotEncoder()

In [44]:
pipeline_workflow = make_pipeline( LinearRegression())

In [45]:
pipeline_workflow

Pipeline(steps=[('linearregression', LinearRegression())])

In [46]:
pipeline_workflow.fit(X_train,y_train)

Pipeline(steps=[('linearregression', LinearRegression())])

In [47]:
pipeline_workflow.score(X_train,y_train)

0.9875704594531879

In [48]:
model = pipeline_workflow.fit(X,y)
model.fit(X,y)

Pipeline(steps=[('linearregression', LinearRegression())])

In [49]:
pipeline_workflow.predict(X_test)

array([0.87160894, 0.43076883, 3.2339119 , ..., 5.2727369 , 5.58047574,
       2.97618937])

#### CREATION DE L'API : TRANSFERT VIA VSCODE EN FICHIER .PY

In [50]:
X.mean()

total_air            87.804083
indoor_pollution     50.481504
outdoor_particule    35.114982
dtype: float64

In [56]:
from fastapi import FastAPI
import pickle
from typing import Union
import numpy as np

app = FastAPI()


@app.get("/")
async def root(total_air: Union[float, None] = 87.8,
               indoor_pollution: Union[float, None] = 50.48,
               outdoor_particule: Union[float, None] = 35.11):
    
    x_user = get_user_df(total_air, indoor_pollution , outdoor_particule)
    pickle_model = pickle.load(open("pipeline_air_pollution.pkl","rb"))  
    pickle_predict = pickle_model.predict(x_user)
    return {'predict': f"{pickle_predict[0]}"}


def get_user_df(total_air, indoor_pollution , outdoor_particule):
    x_user = pd.DataFrame({ 
    "total_air" : [float(total_air)],
    "indoor_pollution" : [float(indoor_pollution)],
    "outdoor_particule" : [float(outdoor_particule)]})
    return x_user

In [57]:
pickle.dump(model, open("pipeline_air_pollution.pkl", "wb"))
pickle_model = pickle.load(open("pipeline_air_pollution.pkl","rb"))

In [60]:
def get_user_df(total_air, indoor_pollution , outdoor_particule):
    x_user = pd.DataFrame({ 
    "total_air" : [float(total_air)],
    "indoor_pollution" : [float(indoor_pollution)],
    "outdoor_particule" : [float(outdoor_particule)]})
    return x_user

#### TEST DU MODEL DE PREDICTION

In [61]:
pickle_model.predict(get_user_df(250.47, 44.3, 30.44))

array([294.67520718])

In [63]:
get_user_df(19.58, 1.60, 17.6)

total_air  indoor_pollution  outdoor_particule
0      19.58               1.6               17.6

In [64]:
y_pred = pipeline_workflow.predict(X_test)

In [65]:
y_pred

array([0.87160894, 0.43076883, 3.2339119 , ..., 5.2727369 , 5.58047574,
       2.97618937])